File: CANON_SDK/EDSDK_v13.19.0_Macintosh/Sample/Swift/CameraControl/CameraControl/SetRemoteShootingCommand.swift
/******************************************************************************
 *                                                                             *
 *   PROJECT : Eos Digital camera Software Development Kit EDSDK               *
 *                                                                             *
 *   Description: This is the Sample code to show the usage of EDSDK.          *
 *                                                                             *
 *                                                                             *
 *******************************************************************************
 *                                                                             *
 *   Written and developed by Canon Inc.                                       *
 *   Copyright Canon Inc. 2018 All Rights Reserved                             *
 *                                                                             *
 *******************************************************************************/

import Cocoa

class SetRemoteShootingCommand: Command {
    
    fileprivate var _parameter: EdsInt32
    fileprivate var _model: CameraModel
    
    init(model: CameraModel, parameter: EdsInt32) {
        
        _parameter = parameter
        _model = model
        super.init(model: model)
        
    }
    
    override func execute() -> Bool {
        
        var error: EdsError = EdsUInt32(EDS_ERR_OK)
        var uintData: EdsUInt32 = 0
        
        error = EdsGetPropertyData(_model.getCameraObject(), EdsPropertyID(kEdsPropID_LensBarrelStatus), 0, EdsUInt32(MemoryLayout<EdsUInt32>.size), &uintData)
        if (error != EdsUInt32(EDS_ERR_OK) || uintData == 1)
        {
            return true
        }
        
        // Set property
        if (error == EdsUInt32(EDS_ERR_OK))
        {
            error = EdsSendCommand(_model.getCameraObject(), EdsCameraCommand(kEdsCameraCommand_SetRemoteShootingMode), _parameter)
        }
        
        //Notification of error
        let viewNotification = ViewNotification()
        _ = viewNotification.errorNotification(error)
        return true
        
    }
}

File: src/CANON_SDK/EDSDK_v13.19.0_Windows/sample/CSharp/CameraControl/CameraControl/Command/SetRemoteShootingCommand.cs
﻿/******************************************************************************
*                                                                             *
*   PROJECT : Eos Digital camera Software Development Kit EDSDK               *
*                                                                             *
*   Description: This is the Sample code to show the usage of EDSDK.          *
*                                                                             *
*                                                                             *
*******************************************************************************
*                                                                             *
*   Written and developed by Canon Inc.                                       *
*   Copyright Canon Inc. 2018 All Rights Reserved                             *
*                                                                             *
*******************************************************************************/

using System;

namespace CameraControl
{
    class SetRemoteShootingCommand : Command
    {
        private uint _parameter;

        public SetRemoteShootingCommand(ref CameraModel model, uint parameter) : base(ref model)
        {
            _parameter = parameter;
        }

        // Execute command	
        public override bool Execute()
        {
            uint outPropertyData;
            uint err = EDSDKLib.EDSDK.EdsGetPropertyData(_model.Camera, EDSDKLib.EDSDK.PropID_LensBarrelStatus, 0, out outPropertyData);
            if (err != EDSDKLib.EDSDK.EDS_ERR_OK || outPropertyData == (uint)EDSDKLib.EDSDK.DcRemoteShootingMode.DcRemoteShootingModeStart)
            //if (err != EDSDKLib.EDSDK.EDS_ERR_OK || outPropertyData == _parameter)
            {
                return true;
            }

            err = EDSDKLib.EDSDK.EdsSendCommand(_model.Camera, EDSDKLib.EDSDK.CameraCommand_SetRemoteShootingMode, (int)_parameter);

            //Notification of error
            if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
            {
                // It retries it at device busy
                if (err == EDSDKLib.EDSDK.EDS_ERR_DEVICE_BUSY)
                {
                    CameraEvent e = new CameraEvent(CameraEvent.Type.DEVICE_BUSY, IntPtr.Zero);
                    _model.NotifyObservers(e);
                    return true;
                }
                else
                {
                    CameraEvent e = new CameraEvent(CameraEvent.Type.ERROR, (IntPtr)err);
                    _model.NotifyObservers(e);
                    return true;
                }
            }

            return true;
        }
    }
}
